### PR TITLE
[otbn] Initial DIF and unit test

### PR DESCRIPTION
**Pull request by imphil**
An initial DIF implementation for OTBN.

Signed-off-by: Philipp Wagner <phw@lowrisc.org>

### CODE REVIEW

**Comment by imphil** [comment]
This DIF isn't ready yet (neither complete nor polished), but it works and highlights the general interface and some potential design decisions that need to be taken care of.

File: sw/device/lib/dif/dif_otbn.h [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+ * @param otbn OTBN instance
+ * @param[out] busy OTBN is busy
+ * @return `kDifOtbnBadArg` if `otbn` is `NULL`, `kDifOtbnOk` otherwise.
+ */
+dif_otbn_result_t dif_otbn_is_busy(const dif_otbn_t *otbn, bool *busy);
+
+// TODO: Think about making src/len in the imem/dmem write/read functions words
+// instead of bytes. Easy for imem with 32b words, harder for dmem with WLEN
+// words. And ideally, we have them consistent: either both are words, or none
+// of them are words.
```
Comment (imphil) [inline sw/device/lib/dif/dif_otbn.h:111]: I know other DIFs were pushed to use something like uint32_t if only word writes are supported. That would be possible for the IMEM size of OTBN, but the DMEM side would need uint256_t. Something to be figured out.

**Comment by lenary** [comment]
Is this ready for design review by @gkelly and me?

File: sw/device/lib/dif/meson.build [hunk 2]
```diff
@@ -69,3 +69,17 @@ sw_lib_dif_i2c = declare_dependency(
     ],
   )
 )
+
+# OTBN DIF library
+sw_lib_dif_otbn = declare_dependency(
+  link_with: static_library(
+    'sw_lib_dif_otbn',
+    sources: [
+      hw_ip_otbn_reg_h,
```
Comment (lenary) [inline sw/device/lib/dif/meson.build:78]: These are defined in the top-level `meson.build` - adding a line there will solve your issue currently seen in CI.

Comment (imphil) [reply sw/device/lib/dif/meson.build:78]: It's coming in as part of another PR, don't worry about build failures at this point.

**Comment by imphil** [comment]
I'd appreciate an initial look if things are going in the right direction, but there are clearly many rough edges that are currently being worked on. Probably most important is the question about the IMEM/DMEM loading data type.

**Comment by imphil** [comment]
@mcy did have a look at the DIF in https://github.com/lowRISC/opentitan/pull/2631, where a version of the DIF appeared as part of a development branch. I'll address those comments here when I look at the DIF next time.

**Comment by imphil** [comment]
Most review comments from @mcy in  #2631 have been included in this PR now. The comments around words vs. bytes in the API, and the right granularity of return codes remains an open where more input is needed.

File: sw/device/lib/dif/dif_otbn.h [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+ */
+dif_otbn_result_t dif_otbn_reset(const dif_otbn_t *otbn);
+
+/**
+ * Start the execution of the application loaded into OTBN.
+ *
+ * @param otbn OTBN instance
+ * @return `kDifOtbnBadArg` if `otbn` is `NULL`, `kDifOtbnOk` otherwise.
+ */
+dif_otbn_result_t dif_otbn_start(const dif_otbn_t *otbn);
```
Comment (lenary) [inline sw/device/lib/dif/dif_otbn.h:86]: This should take the offset for `start_addr` as a parameter as well (and set it before setting START in the cmd register). If you want this to be word aligned, ensure that is documented.

File: sw/device/lib/dif/dif_otbn.h (hunk 1 above)
Comment (lenary) [reply sw/device/lib/dif/dif_otbn.h:100]: We think these types should reflect the alignment required on the ibex side - as ibex can only do a 32bit load/store at a time, these should be `uint32_t*` if you want them to be at least word-aligned. We're not sure that the ibex-side pointers should be more aligned than 32-bit, as that seems to waste space when the ibex-size accesses are all 32-bits wide maximum. This is especially the case as we think the OTBN peripheral doesn't have the ability to take a single 256-bit write.

By all means document and enforce stricter alignment requirements on the `offset` arguments.

File: sw/device/lib/dif/dif_otbn.h [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+ * @return `kDifOtbnBadArg` if `otbn` is `NULL`, `kDifOtbnOk` otherwise.
+ */
+dif_otbn_result_t dif_otbn_is_busy(const dif_otbn_t *otbn, bool *busy);
+
+// TODO: Think about making src/len in the imem/dmem write/read functions words
+// instead of bytes. Easy for imem with 32b words, harder for dmem with WLEN
+// words. And ideally, we have them consistent: either both are words, or none
+// of them are words.
+
+// TODO: We probably need more fine-granular error codes.
```
Comment (lenary) [inline sw/device/lib/dif/dif_otbn.h:102]: We're delaying feedback on this until you have a concrete suggestion.

**Comment by imphil** [comment]
This PR is now ready for review. I went ahead and added the unit test to it as well.

@lenary @gkelly please have a look.

**Review by lenary — APPROVED** [review]
I'm happy with the DIF. If there are issues, we'll catch them in a follow-up PR.

File: sw/device/lib/dif/dif_otbn.c [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+  }
+}
+
+dif_otbn_result_t dif_otbn_imem_write(const dif_otbn_t *otbn,
+                                      uint32_t offset_bytes,
+                                      const uint32_t *src, size_t len_bytes) {
+  // Only 32b-aligned 32b word accesses are allowed.
+  if (otbn == NULL || src == NULL || len_bytes % 4 != 0 ||
+      offset_bytes % 4 != 0 ||
+      offset_bytes + len_bytes > OTBN_IMEM_SIZE_BYTES) {
```
Comment (gkelly) [inline sw/device/lib/dif/dif_otbn.c:234]: I was originally writing a comment about how I didn't like `offset_bytes` and `len_bytes`, but I think I've convinced myself that callers doing:

```c
static const uint32_t kProgram[] = { ... };
const dif_otbn_result_t result = dif_otbn_imem_write(otbn, offset, kProgram, sizeof(kProgram));
```

Is probably the common pattern, and every call site doing a `sizeof(kProgram) / sizeof(uint32_t)` is more fraught than letting the `_write`/`_read` functions take care of it.

File: sw/device/lib/dif/dif_otbn.h [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+   * This error is recoverable and the operation can be retried after correcting
+   * the problem with the argument(s).
+   */
+  kDifOtbnBadArg,
+
+  /**
+   * The hardware returned unexpected data.
+   *
+   * This error either indicates malfunctioning hardware, or a mismatch between
+   * software and hardware.
```
[the remark this answers is not in the document]
Comment (imphil) [reply sw/device/lib/dif/dif_otbn.h:70]: The currently only place where this value is returned is at `dif_otbn_get_err_code()`. I can imagine software simply assuming kDifOtbnUnexpectedData == "unknown error" and continuing. It really depends on the caller.

File: sw/device/lib/dif/dif_otbn.h [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+  kDifOtbnUnexpectedData,
+} dif_otbn_result_t;
+
+/**
+ * Error codes received from the hardware.
+ */
+typedef enum dif_otbn_err_code {
+  // Keep error codes in sync with the hardware!
+  /** No error occurred. */
+  kDifOtbnErrorCodeNoError = 0x0,
```
[the remark this answers is not in the document]
Comment (imphil) [reply sw/device/lib/dif/dif_otbn.h:81]: Unfortunately, the constants as hardware uses them are only available in human-readable text form in the spec, but not part of the hjson file, or the generated register definitions. That's the case for all IP blocks currently. I therefore tried to catch all errors we can possibly catch in the DIF, but ultimately only generating these definitions from a single source will catch all potential issues. (E.g. changing the semantic meaning of error `1`.)

File: sw/device/lib/dif/dif_otbn.c (hunk 5 above)
Comment (imphil) [reply sw/device/lib/dif/dif_otbn.c:234]: Yeah, no fantastic solutions here. The main challenge is consistency between imem and dmem accesses, which have different word width (and hence size and alignment requirements). Having imem_write specify the size in 32b words, where dmem_write uses 256b words, feels very fragile and easy to get wrong.

**Pull request merged**